pkg/mover/intermediate.go

```diff
@@ -119,9 +129,113 @@ func tarImages(imageChanges []*internal.ImageChange, logger Logger) (string, err
 		logger.Printf("Processing image %s\n", change.ImageReference.Name())
 	}
 
-	logger.Printf("Writing all %d images...\n", len(refToImage))
+	logger.Printf("Writing %d images...\n", len(refToImage))
 	if err := tarball.MultiRefWrite(refToImage, imagesFile); err != nil {
 		return "", err
 	}
 	return imagesFile.Name(), nil
 }
+
+// IsIntermediateBundle returns tue only if VerifyIntermediateBundle finds no errors
+func IsIntermediateBundle(bundlePath string) bool {
+	return VerifyIntermediateBundle(bundlePath) == nil
+}
+
+type fileValidations struct {
+	filename, format string
+	validate         func(io.Reader) error
+}
+
+// VerifyIntermediateBundle returns true if the path points to an uncompressed
+// tarball with:
+//  A hints.yaml YAML file
+//  A manifest.json for the images
+//  A directory container an unpacked chart directory with valid YAMLs Chart.yaml & values.yaml
+func VerifyIntermediateBundle(bundlePath string) error {
+	validations := []fileValidations{
+		{filename: "hints.yaml", format: "YAML", validate: validateYAML},
+		{filename: originalChart + "/Chart.yaml", format: "YAML", validate: validateYAML},
+		{filename: originalChart + "/values.yaml", format: "YAML", validate: validateYAML},
+		{filename: imagesTar, format: "TAR", validate: validateTAR},
+	}
+	for _, fv := range validations {
+		r, err := openFromTar(bundlePath, fv.filename)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s from tar: %w", fv.filename, err)
+		}
+		defer r.Close()
+		if err := fv.validate(r); err != nil {
+			return fmt.Errorf("%w: %s is not valid %s: %v",
+				ErrNotIntermediateBundle, fv.filename, fv.format, err)
+		}
+	}
+	return nil
+}
+
+type intermediateBundle struct {
+	bundlePath string
+}
+
+func openBundle(bundlePath string) (*intermediateBundle, error) {
+	if err := VerifyIntermediateBundle(bundlePath); err != nil {
+		return nil, err
+	}
+	return &intermediateBundle{bundlePath}, nil
+}
+
+func (ib *intermediateBundle) ExtractChartTo(dir string) error {
+	err := untar(ib.bundlePath, originalChart, dir)
+	if err != nil {
+		return fmt.Errorf("failed to untar chart from bundle %s into %s: %w",
+			ib.bundlePath, dir, err)
+	}
+	return nil
+}
+
+func (ib *intermediateBundle) LoadHints(log Logger) ([]byte, error) {
+	r, err := openFromTar(ib.bundlePath, IntermediateBundleHintsFilename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract %s from bundle at %s: %w",
+			IntermediateBundleHintsFilename, ib.bundlePath, err)
+	}
+	return io.ReadAll(r)
+}
+
+func refToTag(imageRef name.Reference) (name.Tag, error) {
```

Comment:
Would you care explaining what this does?

Reply:
will add a comment

Reply:
as agreed on the call I will also keep only the cast, removing "manual tag rebuilding code".